OptimizelySDK/DecisionService.cs

```diff
@@ -0,0 +1,331 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using OptimizelySDK.Entity;
+using OptimizelySDK.ErrorHandler;
+using OptimizelySDK.Logger;
+using OptimizelySDK.Utils;
+
+namespace OptimizelySDK
+{
+    /// <summary>
+    /// Optimizely's decision service that determines which variation of an experiment the user will be allocated to.
+    /// The decision service contains all logic around how a user decision is made.  
+    /// This includes the following:
+    /// 1. Checking experiment status
+    /// 2. Checking whitelisting
+    /// 3. Checking sticky bucketing
+    /// 4. Checking audience targeting
+    /// 5. Using Murmurhash3 to bucket the user.
+    /// </summary>
+    public class DecisionService
+    {
+        public enum DecisionType
+        {
+            Unknown,
+            ExperimentNotRunning,
+            Whitelisting,
+            StickyBucketing,
+            AudienceTargeting
+        }
+
+        private Bucketer Bucketer;
+        private IErrorHandler ErrorHandler;
+        private ProjectConfig ProjectConfig;
+        private UserProfileService UserProfileService;
+        private ILogger Logger;
+        
+        /// <summary>
+        ///  Initialize a decision service for the Optimizely client.
+        /// </summary>
+        /// <param name = "bucketer" > Base bucketer to allocate new users to an experiment.</param>
+        /// <param name = "errorHandler" > The error handler of the Optimizely client.</param>
+        /// <param name = "projectConfig" > Optimizely Project Config representing the datafile.</param>
+        /// <param name = "userProfileService" ></ param >
+        /// < param name= "logger" > UserProfileService implementation for storing user info.</param>
+        public DecisionService(Bucketer bucketer, IErrorHandler errorHandler, ProjectConfig projectConfig, UserProfileService userProfileService, ILogger logger)
+        {
+            Bucketer = bucketer;
+            ErrorHandler = errorHandler;
+            ProjectConfig = projectConfig;
+            UserProfileService = userProfileService;
+            Logger = logger;
+        }
+
+        /// <summary>
+        /// Get a Variation of an Experiment for a user to be allocated into.
+        /// </summary>
+        /// <param name = "experiment" > The Experiment the user will be bucketed into.</param>
+        /// <param name = "userId" > The userId of the user.
+        /// <param name = "filteredAttributes" > The user's attributes. This should be filtered to just attributes in the Datafile.</param>
+        /// <returns>The Variation the user is allocated into.</returns>
+        public Variation GetVariation(Experiment experiment, string userId, UserAttributes filteredAttributes)
+        {
+            switch (GetDecisionType(experiment, userId, filteredAttributes))
```

Comment:
This is a departure of how the logic flows through in our other SDKs. The addition of this `GetDecisionType` can confuse other team members who might be working on this from our end. I'd prefer that we stick to the logic we have in the rest of the SDKs: https://github.com/optimizely/php-sdk/blob/master/src/Optimizely/DecisionService/DecisionService.php#L90